src/main/java/org/opentripplanner/netex/mapping/calendar/DayTypeAssignmentMapper.java

```diff
@@ -145,24 +148,58 @@ private void addSpecificDate(DayTypeAssignment dayTypeAssignment) {
     addDate(isDayTypeAvailableForAssigment(dayTypeAssignment), dayTypeAssignment.getDate());
   }
 
-  private void addOperationPeriod(DayTypeAssignment dayTypeAssignment) {
+  private void addOperatingPeriod(DayTypeAssignment dayTypeAssignment) {
     boolean isAvailable = isDayTypeAvailableForAssigment(dayTypeAssignment);
 
     String ref = dayTypeAssignment.getOperatingPeriodRef().getRef();
-    OperatingPeriod period = operatingPeriods.lookup(ref);
+    OperatingPeriod_VersionStructure period = operatingPeriods.lookup(ref);
 
     if (period != null) {
-      Set<DayOfWeek> daysOfWeek = daysOfWeekForDayType(dayType);
+      if (period instanceof OperatingPeriod operatingPeriod) {
+        Set<DayOfWeek> daysOfWeek = daysOfWeekForDayType(dayType);
+        // Plus 1 to make the end date exclusive - simplify the loop test
+        LocalDateTime endDate = operatingPeriod.getToDate().plusDays(1);
+        LocalDateTime date = operatingPeriod.getFromDate();
+
+        addDates(isAvailable, daysOfWeek, endDate, date);
+      } else if (period instanceof UicOperatingPeriod uicOperatingPeriod) {
+        LocalDateTime endDate = uicOperatingPeriod.getToDate().plusDays(1);
+        LocalDateTime date = uicOperatingPeriod.getFromDate();
+
+        addDates(uicOperatingPeriod.getValidDayBits(), isAvailable, endDate, date);
+      }
+    }
+  }
 
-      // Plus 1 to make the end date exclusive - simplify the loop test
-      LocalDateTime endDate = period.getToDate().plusDays(1);
-      LocalDateTime date = period.getFromDate();
+  private void addDates(
+    boolean isAvailable,
+    Set<DayOfWeek> daysOfWeek,
+    LocalDateTime endDate,
+    LocalDateTime date
+  ) {
+    for (; date.isBefore(endDate); date = date.plusDays(1)) {
+      if (daysOfWeek.contains(date.getDayOfWeek())) {
+        addDate(isAvailable, date);
+      }
+    }
+  }
 
-      for (; date.isBefore(endDate); date = date.plusDays(1)) {
-        if (daysOfWeek.contains(date.getDayOfWeek())) {
-          addDate(isAvailable, date);
-        }
+  private void addDates(
+    String validDayBits,
+    boolean isAvailable,
+    LocalDateTime endDate,
+    LocalDateTime date
+  ) {
+    int i = 0;
+    for (; date.isBefore(endDate); date = date.plusDays(1)) {
+      if (i >= validDayBits.length() || validDayBits.charAt(i) == '1') {
+        addDate(isAvailable, date);
+      } else if (validDayBits.charAt(i) != '0') {
+        throw new IllegalArgumentException(
+          "Invalid character '" + validDayBits.charAt(i) + "' in validDayBits"
+        );
       }
+      i++;
```

Comment:
This could be simplified a bit by using the index as the loop condition following?

```suggestion
    for (int i = 0; i < validDayBits.length(); i++) {
      if (i >= validDayBits.length() || validDayBits.charAt(i) == '1') {
        addDate(isAvailable, date.plusDays(1));
      } else if (validDayBits.charAt(i) != '0') {
        throw new IllegalArgumentException(
          "Invalid character '" + validDayBits.charAt(i) + "' in validDayBits"
        );
      }
```

or

```suggestion
    for (int i = 0; i < validDayBits.length(); i++) {
      switch (validDayBits.charAt(i)) {
        case '1' -> addDate(isAvailable, date.plusDays(i));
        case '0' -> {}
        default -> throw new IllegalArgumentException(
          "Invalid character '" + validDayBits.charAt(i) + "' in validDayBits"
        );
      }
    }
```

Reply:
We'll improve on the loop conditions. However, we need to iterate over the dates since the validDayBit String can be shorter than the actual period. It defaults to valid for the corresponding dates